scvi/dataset/dataset.py

```diff
@@ -1468,6 +1476,105 @@ def get_batch_mask_cell_measurement(self, attribute_name: str):
         return batch_mask
 
 
+def poisson_gene_selection(
+    X,
+    n_top_genes: int = 4000,
+    use_cuda = True,
+    n_samples: int = 10000,
+    silent : bool = False,
+    batch_size : int = 5000,
+    **kwargs
+):
+    """ Rank and select genes based on the enrichment of zero counts in data
+        compared to a Poisson count model.
+
+        This is based on M3Drop: https://github.com/tallulandrews/M3Drop
+
+        The method accounts for library size internally, a raw count matrix should be provided.
+
+        Instead of Z-test, enrichment of zeros is quantified by posterior
+        probabilites from a binomial model, computed through sampling.
+
+        :param X: Count matrix, preferably a sparse matrix.
+        :param n_top_genes: How many variable genes to select.
+        :param use_cuda: Default: ``False``.
+        :param n_samples: The number of Binomial samples to use to estimate posterior probability
+            of enrichment of zeros for each gene. Default: ``10000``.
+        :param silent: If ``True``, disables the progress bar. Default ``False``.
+        :param batch_size: Size of temporary matrix for incremental calculation. Larger
+            is faster but requires more RAM or GPU memory. (The default should be fine unless
+            there are hundreds of millions cells or millions of genes.) Default ``5000``.
+
+        :return: A ``pd.DataFrame`` with per-gene statistics.
+
+    """
+    from tqdm import tqdm
+
+    use_cuda = use_cuda and torch.cuda.is_available()
+    dev = torch.device('cuda:0' if use_cuda else 'cpu')
+
+    # Make sure were are working with a sparse CSC matrix.
+    from scipy import sparse
+    X = sparse.csc_matrix(X)
+
+    # Calculate empirical statistics.
+    scaled_means = torch.from_numpy(X.sum(0) / X.sum())[0].to(dev)
+    total_counts = torch.from_numpy(X.sum(1))[:, 0].to(dev)
+
+    observed_fraction_zeros = torch.from_numpy(1. - (X > 0).sum(0) / X.shape[0])[0].to(dev)
+
+    # Calculate probability of zero for a Poisson model.
+    # Perform in batches to save memory.
+    n_batches = total_counts.shape[0] // batch_size
+
+    expected_fraction_zeros = torch.zeros(scaled_means.shape).to(dev)
+
+    for i in range(n_batches):
+        total_counts_batch = total_counts[i * batch_size:(i + 1) * batch_size]
+        # Use einsum for outer product.
+        expected_fraction_zeros += torch.exp(-torch.einsum('i,j->ij', [scaled_means, total_counts_batch])).sum(1)
```

Comment:
@vals I'm realizing now... is this just a sum across axis1 of the full outer product? If so, couldn't you just multiply the column vector by the sum of the row vector to get the final result? If so, we don't have to worry about numba, etc.

Reply:
Hey Adam, I haven't had time to get to the other points yet, but I did a benchmark with numba and it was a fair bit slower even when cached. (~1 minute vs ~3 seconds for PyTorch with CPU)

The problem is I need to exponentiate before I sum (if I understand what you mean right). I was trying to find if there was some "opposite" function of `logsumexp` but didn't find anything.

Reply:
ah ok, perhaps what you have is fine then. Sorry for the sidetracking I didn't see the exp^{-x}.